source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
-  if (diagnostic) {
-    *diagnostic = nullptr;
-    libspirv::UseDiagnosticAsMessageConsumer(&hijack_context, diagnostic);
-  }
-
-  size_t total_word_count = 0u;
-  uint32_t id_upper_bound = 0u;
-  spv_result_t ret;
-  uint32_t max_version = 0u;
-
-  // As we need to bump IDs in each module, so as they do not conflict as we
-  // merge them together, we need to make a copy which we can modify.
-  std::vector<std::vector<uint32_t>> tmp_binaries(num_binaries);
-
-  const auto BumpIDs = [&id_upper_bound](uint32_t* binary, size_t word_offset){
-    spv::Id old_id = static_cast<spv::Id>(GetWord(binary, word_offset));
-    SetWord(binary, word_offset, id_upper_bound + old_id);
-  };
-
-  spv_position_t position = {};
-  position.index = 1u;
-  for (uint32_t i = 0u; i < num_binaries; ++i) {
-    const spv_const_binary_t& spvbinary = binaries[i];
-    const uint32_t* binary = spvbinary.code;
-    const size_t word_count = spvbinary.wordCount;
 
-    const uint32_t version = GetWord(binary, 1u);
-    if (version > spv::Version) {
-      return libspirv::DiagnosticStream(position, hijack_context.consumer,
-                                        SPV_ERROR_INVALID_DATA)
-             << "Binary " << i << " uses an unsupported SPIR-V version:"
-             << ((version >> 16u) & 0xff) << "." << ((version >> 8u) & 0xff);
+  // Remove duplicate capabilities
+  std::unordered_set<uint32_t> capabilities;
+  for (auto i = linkedModule->capability_begin(); i != linkedModule->capability_end();) {
+    auto res = capabilities.insert(i->GetSingleWordOperand(0u));
+    if (res.second) {
+      ++i;
+      continue;
     }
-
-    if (version > max_version)
-      max_version = version;
-
-    auto &tmp_binary = tmp_binaries[i];
-    tmp_binary.resize(word_count);
-    std::memcpy(tmp_binary.data(), binary, word_count * sizeof(spirv_word));
-
-    uint32_t binary_max_id = GetWord(binary, 3u) - 1u;
-
-    spv_binary_t tmp_spvbinary = { tmp_binary.data(), word_count };
-    ret = TransformBinary2(hijack_context, tmp_spvbinary, BumpIDs);
-    if (ret != SPV_SUCCESS)
-      return ret;
-
-    id_upper_bound += binary_max_id;
-    total_word_count += tmp_spvbinary.wordCount;
+    i = i.Erase();
   }
 
-  // id_upper_bound is currently equal to the highest id being used, so add
-  // one to get the strict upper bound
-  ++id_upper_bound;
-
-  const bool create_library = (options & SPV_LINKER_OPTION_CREATE_LIBRARY) != 0u;
-
-  *linked_binary = new spv_binary_t;
-  (*linked_binary)->code = new uint32_t[total_word_count];
-
-  ret = MergeModules(hijack_context, tmp_binaries, num_binaries,
-                            id_upper_bound, max_version, **linked_binary,
-                            create_library);
-  if (ret != SPV_SUCCESS) {
-    spvBinaryDestroy(*linked_binary);
-    return ret;
+  // Remove import linkage attributes
+  for (auto i = linkedModule->annotation_begin(); i != linkedModule->annotation_end();) {
+    if (i->opcode() != SpvOpDecorate ||
```

Comment:
If the import was not resolved, we wouldn’t be here as the code exits [here](https://github.com/KhronosGroup/SPIRV-Tools/pull/693/files/a45b072e2d85afb1193f8d9e8f879e4a5792b921..5be189f9291d60751246ede2f3b27400a14e151f#diff-202f9912f4c3d52d3531ffa6c2cbae74R170)